page_group.php

```diff
@@ -0,0 +1,145 @@
+<?php
+// This file is part of Moodle - https://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Page group meta data.
+ *
+ * @package   tool_excimer
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use core\chart_line;
+use core\chart_series;
+use tool_excimer\helper;
+use tool_excimer\output\tabs;
+use tool_excimer\page_group;
+
+require_once('../../../config.php');
+require_once($CFG->libdir.'/adminlib.php');
+
+$expiry = get_config('tool_excimer', 'expiry_fuzzy_counts');
+$pagegroupid = required_param('id', PARAM_INT);
+$monthstodisplay = optional_param('monthstodisplay', $expiry, PARAM_INT);
+
+$url = new \moodle_url('/admin/tool/excimer/page_group.php');
+$context = context_system::instance();
+
+$pagegroup = new page_group($pagegroupid);
+
+$PAGE->set_context($context);
+$PAGE->set_url($url);
+$PAGE->navbar->add($pagegroup->get('name'));
+
+admin_externalpage_setup('tool_excimer_report_page_groups');
+
+$output = $PAGE->get_renderer('tool_excimer');
+
+$data = $pagegroup->to_record();
+
+// Data for table.
+$data->month = userdate(strtotime($data->month . '01'), '%b %Y');
+$data->approxcount = pow(2, $data->fuzzycount - 1) . ' - ' . pow(2, $data->fuzzycount);
+$data->approxduration = pow(2, $data->fuzzydurationsum);
+$data->histogram = helper::make_histogram($data);
+
+// Data for charts.
+// Each duration range forms its own line on the chart.
+
+$firstmonth = (int) userdate(strtotime("$monthstodisplay months ago"), '%Y%m'); // YYYYMM format.
```

Comment:
Could we move the handling of this to a helper function? It seems to be used in multiple places, and would be good to avoid the manual cast and format setting.

e.g. `function get_year_month($timestamp): int` or similar.

Places this is used:
- page_group.php
- tests/tool_excimer_purge_page_group_test.php
- classes/task/purge_page_groups.php

Reply:
A class 'monthint' was created to handle logic relating to this value. Feel free to suggest a better name.

Reply:
I hope to look into it a bit more tomorrow, but quickly going over it, do you think it might be better served as a proper OO class? e.g. so you can do things like the below which might be useful come the pagination task:
- `$month->prev()->prev()->formatted()`
- `$month->prev()->prev()->timestamp()`
- `$month->prev()->prev()->month()`

I've omited the typical `get_` prefix but you get the point :)